### PR TITLE
NDRS-529: joiner reactor api endpoints

### PR DESCRIPTION
**Pull request by EdHastingsCasperLabs**
This PR splits up the api_server into more specialized components and wires up the joiner reactor to support REST status, metrics, and the event stream.

https://casperlabs.atlassian.net/browse/NDRS-529

### CODE REVIEW

File: node/src/components/event_stream_server.rs [hunk 1]
```diff
@@ -0,0 +1,108 @@
+//! API server
```
Comment (dwerner) [inline node/src/components/event_stream_server.rs:1]: I think these docs are out of date now that this file is `event_stream_server.rs`

File: node/src/components/rest_server.rs [hunk 2]
```diff
@@ -0,0 +1,149 @@
+//! API server
```
Comment (dwerner) [inline node/src/components/rest_server.rs:1]: Similar to my last comment - I think these docs are out of date now that we are `rest_server.rs`

File: resources/production/config-example.toml [hunk 3]
```diff
@@ -69,20 +69,47 @@ gossip_interval = 120_000
 # only in the unit files themselves via `-C=network.systemd_support=true`.
 systemd_support = false
 
+
 # =============================================
-# Configuration options for the HTTP API server
+# Configuration options for the HTTP RPC server
```
Comment (dwerner) [inline resources/production/config-example.toml:74]: Perhaps `JSON-RPC` here?

File: node/src/reactor/validator/memory_metrics.rs [hunk 4]
```diff
@@ -20,8 +20,8 @@ pub(super) struct MemoryMetrics {
     mem_storage: IntGauge,
     /// Estimated heap memory usage of contract runtime component.
     mem_contract_runtime: IntGauge,
-    /// Estimated heap memory usage of api server component.
-    mem_api_server: IntGauge,
+    /// Estimated heap memory usage of rpc server component.
+    mem_rpc_server: IntGauge,
```
Comment (Fraser999) [inline node/src/reactor/validator/memory_metrics.rs:24]: We don't want to include the REST or event stream servers in here?

**Review by Fraser999 — COMMENTED** [review]
rpc_server.rs could do with having its module-level docs updated (I can't add a suggestion to the PR as it's not marked as changed code in the PR)

Also, the config file at utils/nctl/templates/node-config.toml should be updated with the new sections too.

File: node/src/reactor/validator/memory_metrics.rs (hunk 4 above)
Comment (EdHastingsCasperLabs) [reply node/src/reactor/validator/memory_metrics.rs:24]: Added.

File: node/src/components/rpc_server/http_server.rs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+    let server_with_shutdown = server.with_graceful_shutdown(async {
+        shutdown_receiver.await.ok();
+    });
+
+    let server_joiner = tokio::spawn(server_with_shutdown);
+
+    let _ = server_joiner.await;
+
+    // Kill the event-stream handlers, and shut down the server.
```
[the remark this answers is not in the document]
Comment (EdHastingsCasperLabs) [reply node/src/components/rpc_server/http_server.rs]: Did this.

File: node/src/components/rpc_server/http_server.rs [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+use super::{
+    rpcs::{self, RpcWithOptionalParamsExt, RpcWithParamsExt, RpcWithoutParamsExt},
+    Config, ReactorEventT,
+};
+use crate::{effect::EffectBuilder, utils};
+
+/// Run the HTTP server.
+///
+/// `data_receiver` will provide the server with local events which should then be sent to all
+/// subscribed clients.
```
[the remark this answers is not in the document]
Comment (EdHastingsCasperLabs) [reply node/src/components/rpc_server/http_server.rs]: Did this.

File: node/src/components/rpc_server/config.rs [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+///
+/// Uses a fixed port per node, but binds on any interface.
+const DEFAULT_ADDRESS: &str = "0.0.0.0:7777";
+
+/// API server configuration.
+#[derive(Clone, DataSize, Debug, Deserialize, Serialize)]
+// Disallow unknown fields to ensure config files and command-line overrides contain valid keys.
+#[serde(deny_unknown_fields)]
+pub struct Config {
+    /// Address to bind HTTP server to.
```
[the remark this answers is not in the document]
Comment (EdHastingsCasperLabs) [reply node/src/components/rpc_server/config.rs]: Did this.

File: node/src/components/rpc_server/config.rs [hunk 8]
```diff
@@ -0,0 +1,31 @@
+use datasize::DataSize;
+use serde::{Deserialize, Serialize};
+
+/// Default binding address for the HTTP server.
+///
+/// Uses a fixed port per node, but binds on any interface.
+const DEFAULT_ADDRESS: &str = "0.0.0.0:7777";
+
+/// API server configuration.
```
[the remark this answers is not in the document]
Comment (EdHastingsCasperLabs) [reply node/src/components/rpc_server/config.rs]: Did this.

File: node/src/components/rpc_server/config.rs [hunk 9]
```diff
@@ -0,0 +1,31 @@
+use datasize::DataSize;
+use serde::{Deserialize, Serialize};
+
+/// Default binding address for the HTTP server.
```
[the remark this answers is not in the document]
Comment (EdHastingsCasperLabs) [reply node/src/components/rpc_server/config.rs]: Did this.

File: node/src/components/rest_server/config.rs [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+///
+/// Uses a fixed port per node, but binds on any interface.
+const DEFAULT_ADDRESS: &str = "0.0.0.0:8888";
+
+/// API server configuration.
+#[derive(Clone, DataSize, Debug, Deserialize, Serialize)]
+// Disallow unknown fields to ensure config files and command-line overrides contain valid keys.
+#[serde(deny_unknown_fields)]
+pub struct Config {
+    /// Address to bind HTTP server to.
```
[the remark this answers is not in the document]
Comment (EdHastingsCasperLabs) [reply node/src/components/rest_server/config.rs]: Did this.

File: node/src/components/rest_server/config.rs [hunk 11]
```diff
@@ -0,0 +1,31 @@
+use datasize::DataSize;
+use serde::{Deserialize, Serialize};
+
+/// Default binding address for the HTTP server.
+///
+/// Uses a fixed port per node, but binds on any interface.
+const DEFAULT_ADDRESS: &str = "0.0.0.0:8888";
+
+/// API server configuration.
```
[the remark this answers is not in the document]
Comment (EdHastingsCasperLabs) [reply node/src/components/rest_server/config.rs]: Did this.

File: node/src/components/rest_server/config.rs [hunk 12]
```diff
@@ -0,0 +1,31 @@
+use datasize::DataSize;
+use serde::{Deserialize, Serialize};
+
+/// Default binding address for the HTTP server.
```
[the remark this answers is not in the document]
Comment (EdHastingsCasperLabs) [reply node/src/components/rest_server/config.rs]: Did this.

File: node/src/components/event_stream_server/config.rs [hunk 13]
```diff
@@ -4,13 +4,13 @@ use serde::{Deserialize, Serialize};
 /// Default binding address for the HTTP server.
```
[the remark this answers is not in the document]
Comment (EdHastingsCasperLabs) [reply node/src/components/event_stream_server/config.rs]: Did this.

File: node/src/components/event_stream_server/config.rs [hunk 14]
```diff
@@ -4,13 +4,13 @@ use serde::{Deserialize, Serialize};
...
 ///
 /// Uses a fixed port per node, but binds on any interface.
-const DEFAULT_ADDRESS: &str = "0.0.0.0:7777";
+const DEFAULT_ADDRESS: &str = "0.0.0.0:9999";
 
 /// Default number of SSEs to buffer.
 const DEFAULT_EVENT_STREAM_BUFFER_LENGTH: u32 = 100;
 
 /// API server configuration.
```
[the remark this answers is not in the document]
Comment (EdHastingsCasperLabs) [reply node/src/components/event_stream_server/config.rs]: Did this.

File: node/src/components/rest_server.rs [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+//! JSON-encoded data.
+//!
+//! The actual server is run in backgrounded tasks.   RPCs requests are translated into reactor
+//! requests to various components.
+//!
+//! This module currently provides both halves of what is required for an API server: An abstract
+//! API Server that handles API requests and an external service endpoint based on HTTP.
+//!
+//! For the list of supported RPCs and SSEs, see
+//! https://github.com/CasperLabs/ceps/blob/master/text/0009-client-api.md#rpcs
```
[the remark this answers is not in the document]
Comment (EdHastingsCasperLabs) [reply node/src/components/rest_server.rs]: Did this.

File: node/src/components/event_stream_server/event.rs [hunk 16]
```diff
@@ -0,0 +1,39 @@
+use std::fmt::{self, Display, Formatter};
+
+use derive_more::From;
+
+use crate::types::{
+    json_compatibility::ExecutionResult, BlockHash, BlockHeader, DeployHash, DeployHeader,
+    FinalizedBlock,
+};
+
+#[derive(Debug, From)]
```
[the remark this answers is not in the document]
Comment (EdHastingsCasperLabs) [reply node/src/components/event_stream_server/event.rs]: Did this.

File: node/src/components/event_stream_server/event.rs [hunk 17]
```diff
@@ -0,0 +1,39 @@
+use std::fmt::{self, Display, Formatter};
+
+use derive_more::From;
+
```
[the remark this answers is not in the document]
Comment (EdHastingsCasperLabs) [reply node/src/components/event_stream_server/event.rs]: Did this.

File: node/src/components/event_stream_server.rs [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+//! JSON-encoded data.
+//!
+//! The actual server is run in backgrounded tasks.   RPCs requests are translated into reactor
+//! requests to various components.
+//!
+//! This module currently provides both halves of what is required for an API server: An abstract
+//! API Server that handles API requests and an external service endpoint based on HTTP.
+//!
+//! For the list of supported RPCs and SSEs, see
+//! https://github.com/CasperLabs/ceps/blob/master/text/0009-client-api.md#rpcs
```
[the remark this answers is not in the document]
Comment (EdHastingsCasperLabs) [reply node/src/components/event_stream_server.rs]: Did this.

File: resources/production/config-example.toml (hunk 3 above)
Comment (EdHastingsCasperLabs) [reply resources/production/config-example.toml]: Did this.

File: node/src/components/rest_server.rs (hunk 2 above)
Comment (EdHastingsCasperLabs) [reply node/src/components/rest_server.rs]: Did this.

File: node/src/components/event_stream_server.rs (hunk 1 above)
Comment (EdHastingsCasperLabs) [reply node/src/components/event_stream_server.rs]: Did this.

**Review by dwerner — APPROVED** [review]
:+1:

**Pull request merged**